README.md
# Sabre-Zarafa

The aim of this project is to provide a full CardDav backend for
[SabreDAV](http://http://sabre.io/dav) to connect with
[Zarafa](http://www.zarafa.com) groupware.

Tarballs and zipfiles of the source can be downloaded
[here](https://github.com/1afa/sabre-zarafa/releases). See below for
installation instructions, you will also need to download and install SabreDAV
and log4php. For an overview of the changes, see the
[ChangeLog](https://github.com/1afa/sabre-zarafa/blob/master/ChangeLog).

Sabre-Zarafa is a backend for the SabreDAV server. SabreDAV is a generic DAV
server that processes CardDAV, CalDAV and WebDAV requests. It handles all the
intricate parts of the DAV protocols and client communication, but doesn't know
anything about databases or stores. It defers the responsibility for providing
abstract objects like "cards" and "address books" to backend software like
Sabre-Zarafa, which is free to implement retrieval and storage however it
likes.

Sabre-Zarafa knows nothing about DAV, but does know how to get data from the
Zarafa server and convert it to VCard format. Together, the SabreDAV frontend
and the Sabre-Zarafa backend combine to form a Zarafa-based CardDAV server.

Sabre-Zarafa is pretty useable, but it's not all the way there yet. Patches and
improvements are welcome.

This particular repository is a continuation of the original Sabre-Zarafa
project, which is hosted [at Google Code](http://code.google.com/p/sabre-zarafa),
and which has not been maintained for a while. This repository is not really a
fork, but is intended to be the new mainline tree. Work on Sabre-Zarafa is
sponsored by [1A First Alternative](http://www.1a-first-alternative.com).

## License

Sabre-Zarafa is licensed under the terms of the [GNU Affero General Public
License, version 3](http://www.gnu.org/licenses/agpl-3.0.html).

## Install

### Introduction

This installs as any [SabreDAV](http://sabre.io/dav) server.
Unpack the source into a directory. This readme will assume that
`/var/www/htdocs/sabre-zarafa` is the root directory of the install.

### Download and install Composer

Composer is required to install Sabre-Zarafa's dependencies.
Installation instructions can be found on the
[Composer website](https://getcomposer.org/doc/00-intro.md#installation-nix).

After installation is done, run the following two commands:

    # cd /var/www/htdocs/sabre-zarafa
    # composer install

### Configure permissions and logging

The webserver needs to write to the `data` directory, since it is used by
SabreDAV to store DAV locks. (NB, the author is not convinced that CardDAV
actually uses locking.) The log file, called `debug.txt`, should also be
writable. If your server runs as the user `apache`:

    # chown apache:apache /var/www/htdocs/sabre-zarafa/data
    # chmod 0750 /var/www/htdocs/sabre-zarafa/data
    # chown apache:apache /var/www/htdocs/sabre-zarafa/debug.txt
    # chmod 0640 /var/www/htdocs/sabre-zarafa/debug.txt

Sabre-Zarafa logs using [log4php](http://logging.apache.org/log4php/).
To configure logging you need to edit `log4php.xml`.
Default setup is to log WARN, ERROR and FATAL messages to `debug.txt` with a maximum size of 5MB for logfile and 3 backup indexes (`debug.txt.1`).
If you want to log to some other file, insert the proper filename here.
(The default config file uses `debug.txt` in the application dir, which is probably not what you want.)
To disable debugging simply set the root appender to `noDebug`.
Optionally, choose a max logging level to observe.
In order from less to more logging, you can specify `FATAL`, `ERROR`, `WARN`, `INFO`, `DEBUG`, and `TRACE`.

Example config for logging all errors at or higher than INFO to `/var/log/sabrezarafa.log`:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<configuration xmlns="http://logging.apache.org/log4php/">
	<appender name="fileAppender" class="LoggerAppenderRollingFile">
		<layout class="LoggerLayoutPattern">
			<param name="conversionPattern" value="%d{Y-m-d H:i:s} %-5p %c %x - %m%n" />
		</layout>
		<param name="file" value="/var/log/sabrezarafa.log" />
		<param name="maxFileSize" value="5MB" />
		<param name="maxBackupIndex" value="3" />
	</appender>
	<root>
		<appender_ref ref="fileAppender" />
		<level value="INFO" />
	</root>
</configuration>
```

Example config for suppressing all logging by specifying the null appender:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<configuration xmlns="http://logging.apache.org/log4php/">
	<appender name="noDebug" class="loggerAppenderNull" />
        <root>
		<appender_ref ref="noDebug" />
	</root>
</configuration>
```

### Sabre-Zarafa configuration

Sabre-Zarafa needs additional configuration in `config.inc.php`. You must set
`CARDDAV_ROOT_URI` to the proper value. This is the path from the root of the
webserver to Sabre-Zarafa. If Sabre-Zarafa is installed in the webserver root,
then use `/`. If Sabre-Zarafa is installed in `/var/www/htdocs/sabre-zarafa`
and `/var/www/htdocs` is the server root, then put `/sabre-zarafa`. If you're
not using `mod_redirect` to redirect requests to `server.php`, use
`/sabre-zarafa/server.php`.

You can also adjust other settings, some are highly experimental (non-UTF8
vcards for instance) and should be used only for testing.

If you prefer to only use read operations and not make any edits to the
database, set `READ_ONLY` to `true`.

### Running from the root of the webserver

According to the SabreDAV documentation, you get the least issues if you run
the service straight from the root of the webserver. You can run it on a
standard port like 80 or 443, but for CardDAV, it makes some sense to use port
8843, since that's [what OSX Addressbook uses by
default](http://support.apple.com/kb/ts1629). Always enable SSL in production
environments, since Sabre-Zarafa uses Basic authentication, and that's only
secure when used over an encrypted connection.

To configure Apache to listen to port 8843 and use a virtual host to serve
Sabre-Zarafa, put something like the following configuration in `httpd.conf`:

    Listen 8843

    <VirtualHost *:8843>

        # ...general server options, enable PHP parsing, SSL setup, etc...

        DocumentRoot /var/www/htdocs/sabre-zarafa
        <Directory /var/www/htdocs/sabre-zarafa>
            DirectoryIndex server.php
            RewriteEngine On
            RewriteBase /

            # If the request does not reference an actual plain file or
            # directory (such as server.php), interpret it as a "virtual path"
            # and pass it to server.php:
            RewriteCond %{REQUEST_FILENAME} !-f
            RewriteCond %{REQUEST_FILENAME} !-d
            RewriteRule ^.*$ /server.php

            # If you're getting 412 Precondition Failed errors, try stripping the ETag headers:
            # RequestHeader unset If-None-Match
            # Header unset ETag
            # FileETag None
        </Directory>

        # Files and directories writable by the server should never be public:
        <Directory /var/www/htdocs/sabre-zarafa/data>
            Deny from all
        </Directory>
        <Files /var/www/htdocs/sabre-zarafa/debug.txt>
            Deny from all
        </Files>

        # Deny access to application directories:
        <Directory /var/www/htdocs/sabre-zarafa/lib>
            Deny from all
        </Directory>
        <Directory /var/www/htdocs/sabre-zarafa/vendor>
            Deny from all
        </Directory>
    </VirtualHost>

Don't forget to edit `config.inc.php` and change `CARDDAV_ROOT_URI` to `/`.

### Running from a subdirectory

You can also run Sabre-Zarafa in a subdirectory of your webserver. In that
case, use a variant of this configuration:

    <Directory /var/www/htdocs/sabre-zarafa>
        DirectoryIndex server.php
        RewriteEngine On
        RewriteBase /sabre-zarafa

        # If the request does not reference an actual plain file or directory
        # (such as server.php), interpret it as a "virtual path" and pass it to
        # server.php:
        RewriteCond %{REQUEST_FILENAME} !-f
        RewriteCond %{REQUEST_FILENAME} !-d
        RewriteRule ^.*$ /sabre-zarafa/server.php

        # If you're getting 412 Precondition Failed errors, try stripping the ETag headers:
        # RequestHeader unset If-None-Match
        # Header unset ETag
        # FileETag None
    </Directory>

    # Files and directories writable by the server should never be public:
    <Directory /var/www/htdocs/sabre-zarafa/data>
        Deny from all
    </Directory>
    <Files /var/www/htdocs/sabre-zarafa/debug.txt>
        Deny from all
    </Files>

    # Deny access to application directories:
    <Directory /var/www/htdocs/sabre-zarafa/lib>
        Deny from all
    </Directory>
    <Directory /var/www/htdocs/sabre-zarafa/vendor>
        Deny from all
    </Directory>

Edit `config.inc.php` and change `CARDDAV_ROOT_URI` to `/sabre-zarafa`.

## Testing and troubleshooting

Before testing with a CardDAV client, surf to the Sabre-Zarafa URL with a
browser and check that you can log in and get a listing of your contacts.
Pointing a web browser at Sabre-Zarafa is the easiest way to check that the
install works; only then should you try real CardDAV clients. If things don't
work in the browser, start there first.

If you're not seeing any output on the screen, check your PHP error logs. Maybe
your PHP does not have all the required extensions enabled. SabreDAV uses quite
a few extensions, such as `ctype` and `dom`. Googling any error messages should
shed light on the problem.

If you're still fairly certain that things should be working but they don't,
try changing the error output statements in `server.php` to:

    ini_set('display_errors', true);
    ini_set('html_errors', true);

PHP should now complain loudly when something goes wrong. Also, make sure
`log4php` is installed and enable very verbose logging by setting the log level
to `INFO`, `DEBUG` or even `TRACE` in `log4php.xml`. You will get very chatty
logs that should point out any problems.

Problems with OS X Contacts.app can often be diagnosed by running the program
from a terminal; it prints quite a lot of helpful output when things aren't
working. Type this command in a terminal to launch Contacts.app:

    /Applications/Contacts.app/Contents/MacOS/Contacts


### Filing bugs, getting support

Bugs and issues can be filed at GitHub on the Sabre-Zarafa project page. You'll
need a GitHub account. Alternatively, you can post to the Sabre-Zarafa thread
at the [Zarafa forums](http://forums.zarafa.com); the author checks it regularly.

If you want to request support for certain VCard properties, be sure to include
a traffic dump that shows the exact format of the field.

If your client is having trouble talking to Sabre-Zarafa, please include a
traffic dump that shows the network conversation, if at all possible. Seeing
what goes wrong "on the wire" is indispensable for debugging.

In general, if you get any errors, be as specific as possible about what is
going wrong and how to reproduce the problem. Include as much relevant logs as
possible (both from PHP and Sabre-Zarafa). Bug reports such as "I tried it on a
Mac and it doesn't work" are good to know (the author can perhaps try to
reproduce the problem), but are not directly actionable.

### Authentication

Please note that the authentication backend uses Basic authentication
exclusively. It is not possible to use Digest authentication, because
Sabre-Zarafa needs the user's plaintext password to log in to the Zarafa
server. Some clients will only work with Basic auth if the host uses SSL. You
should *always* enable SSL on the server, since sending plaintext passwords
over an unencrypted connection is a security risk.

Some detailed information about SabreDAV setup are available in [SabreDAV
documentation](http://sabre.io/dav). Do not hesitate to read it!

## Upgrading

### 0.22 to 0.23

Sabre-Zarafa now depends on SabreDAV 2.1. Run `composer update` to update the
dependencies.

### 0.21 to 0.22

Sabre-Zarafa is now installed completely by running
[Composer](http://getcomposer.org) in the toplevel directory. Run `composer
update` to update the dependencies (such as SabreDAV). If you visit the
application in a browser, you'll notice a new design. Internally, the file and
class structure of Sabre-Zarafa received a thorough reworking, but that should
be invisible from the outside.

### 0.20 to 0.21

Sabre-Zarafa 0.21 is written for SabreDAV 1.8.6 and Sabre-VObject 3.0.0 (the
production release, not the alpha or beta versions). You must install at least
those versions of both packages, or newer. Some small API changes have been
introduced between Sabre-VObject 3.0.0-alpha4 and the 3.0.0 release (as well as
many internal fixes), so you must upgrade. The old SabreDAV 1.8.5 can
apparently not handle the Sabre-VObject 3.x branch and must also be updated to
version 1.8.6, which has support for both 2.x and 3.x branches.

This version introduces a new config option, called `FOLDER_RENAME_PATTERN`.
This variable allows you to define a rename pattern for folders. You can define
certain formatting variables; refer to the comments in the config file for
particulars. Updating the config file to set this variable is not required; if
the option is not specified, the previous behaviour is used (folder names are
not rewritten).

### 0.19 to 0.20

Sabre-Zarafa 0.20 is written against the Sabre-VObject 3.x branch, which isn't
yet included in SabreDAV and must be installed manually. Please refer to the
installation instructions above. Because Sabre-VObject 3.x is in its early
days, it's possible that some things don't quite work. Please report any bugs
and issues you find.

### 0.18 to 0.19

Sabre-Zarafa 0.19 adds the `ETAG_ENABLE` config variable to `config.inc.php`.
This variable controls the optional generation of ETags. If Sabre-Zarafa
notices that the variable is not set, it will default to true. So users with an
existing config file don't absolutely need to update their config, though it is
advised.

Sabre-Zarafa 0.19 no longer requires `log4php`; it will run without logging if
`log4php` is not installed. This makes it a little bit easier for new users to
get Sabre-Zarafa up and running (less moving parts).

Other improvements in version 0.19 are all "under the hood" and should require
no special action on the administrator's part.


## PHP < 5.4

As of version 0.22, Sabre-Zarafa requires SabreDAV 2.0, which requires PHP 5.4.

As of version 0.18, Sabre-Zarafa requires SabreDAV 1.8, which in turn requires
PHP 5.3.

## Debugging

### Log4Php

Starting with release 0.10 [log4php](http://logging.apache.org/log4php/) is
used for debugging. Release 0.19 made the installation of log4php optional:
logging is done by the SabreZarafa\Logger class, and if it can't find log4php,
it will silently discard all log messages.


ChangeLog
* Tue May 19 2015 Richard de Vroede <richard@1afa.com> 0.23.1
- Interface incompatibility fixed with sabredav 2.1.3
- Moved repo from github.com/bokxing-it to github.com/1afa for organizational
  reasons

* Wed Dec 10 2014 Alfred Klomp <alfred@1afa.com> 0.23
- Make compatible with SabreDAV 2.1

* Mon Nov 17 2014 Alfred Klomp <alfred@1afa.com> 0.22.3
- VCard\Parser: fix class declaration error (Issue #24)
- VCard\Producer: distinguish between vCard 3 and 4 phone preference formats;
- VCard\Producer: distinguish between vCard 3 and 4 email preference formats;

* Fri Jul 25 2014 Alfred Klomp <alfred@1afa.com> 0.22.2
- Add log configuration instructions to README.md;
- Remove custom Logger class, use stock Log4php singleton everywhere instead;

* Mon Jul 21 2014 Alfred Klomp <alfred@1afa.com> 0.22.1
- Fix showstopper typo in server.php (Issue #21)

* Thu Jul 17 2014 Alfred Klomp <alfred@1afa.com> 0.22
- Move files into a PSR-0 dir structure, load with Composer;
- VCardParser/VCardProducer: rename to Parser/Producer;
- Rename Zarafa_Store class to Store;
- Rename Zarafa_Folder class to Folder;
- Rename Zarafa_Webaccess_Settings class to just WebaccessSettings;
- Rename Zarafa_Bridge class to just Bridge;
- Backends: remove Zarafa_ prefix from classname;
- Place "our" classes in the SabreZarafa namespace;
- Move restrict functions to separate class;
- SabreDAV 2.0: fix runtime error by removing obsolete constant;
- SabreDAV 2.0: update Proppatch handler to the SabreDAV way of doing things;
- Update composer.json to pull in SabreDAV 2.0; also add author names.
- Change FALSE() functions to comparisons with the 'false' constant, to make
  the code more standard.
- Introduce the INCLUDE_SHARED_ADDRESSBOOKS config option to pull in address
  books shared with different users.
- Split off the Webaccess settings into separate class: the ZarafaBridge class
  is getting a bit bloated.
- Add support for shared addressbooks. Thanks to @erwint for donating the core
  logic: https://github.com/erwint/sabre-zarafa/
- ZarafaBridge: add method to get the Zarafa Webaccess settings, which we will
  need shortly in order to find the user's shared folders.
- VCardParser: first collect all email addresses, then sort them in order of
  preference as given by client. Previously the addresses were saved in source
  order, the first three to appear made it to Zarafa.
- FAX,OTHER maps to 'primary_fax_number' and vice versa, which follows the
  convention used by Zarafa Webmail and OSX Addressbook;
- FAX property now maps to 'business_fax_number', which is what Zarafa Webmail
  uses, likewise with TEL,WORK and 'office_telephone_number';
- Fix for a really common bug: avoid PHP fatal error in the folder name
  substitution code by checking if an array value is set before trying to
  access it;

* Wed Jun 26 2013 Alfred Klomp <alfred@bokxing-it.nl> 0.21
- Use shorthand notation to get property parameters; just as good as the long
  version, but way more readable.
- Don't discard values for website and IM when we don't have any MAPI slots
  left; compress their values into a single MAPI slot by making a
  comma-separated list.
- Zarafa_Folder: add support for renaming folders according to a user-defined
  template; this helps in the situation where a public and a private folder
  have the same name.

* Wed Jun 05 2013 Alfred Klomp <alfred@bokxing-it.nl> 0.20
- VCardProducer: change SORT-AS tag to SORT-STRING; SORT-AS comes into play in
  vCard 4.0 as a parameter to the N property; generate it as such.
- VCardParser: add support for many proprietary phone/relation properties.
- Update README.md, add documentation.
- Bugfix: don't display mobile phone number twice.
- VCardParser: add a slew of proprietary instant messaging properties to the IM
  property matcher.
- Migrate the code to Sabre-VObject 3.0.0 format; make backwards-incompatible
  changes that break Sabre-VObject 2.x. The breakage is necessary because
  Sabre-VObject 3.0 breaks compatibility. We need to migrate to Sabre-VObject
  3.0 because it has proper support for escaping multiline fields, such as the
  NOTE: property. These fields break in 2.x.
- Don't parse the REV: tag as a Unix timestamp, and don't fill it with the
  lastmodified timestamp when creating a VCard. Turns out Evolution sets this
  property to a MD5 hash.

* Tue May 28 2013 Alfred Klomp <alfred@bokxing-it.nl> 0.19
- VCardParser and VCardProducer: clean up function interfaces, improve
  encapsulation, split large methods into smaller ones.
- Add ETag support, based on the hash of the EntryID and the last-modified
  timestamp. This should be sufficiently unique. Add a config option to turn
  this feature off if desired. Creating contacts now works in OSX Contacts.
- VCardParser: remove superfluous escaping functions; VObject handles that now;
- VCardParser: rewrite contact photo logic, the old code did not properly
  recognize differences between uppercase and lowercase encoding strings. Also
  build in some more checks, notably the check if the GD extension is loaded
  before trying to use it.
- VCardProducer: don't error out when email display name not found;
- Create separate logging class, which will try to load log4php but silently
  fail if that component is not installed; this loosens the dependency on
  log4php.

* Mon Apr 22 2013 Alfred Klomp <alfred@bokxing-it.nl> 0.18
- Tag version 0.18
- Rewrote email producing code to print pref property;
- Rewrote telephone number parsing code for clarity/solidity;
- Calculate CTag from all cards in directory; this fixes synchronization;
- Add documentation on how to disable ETag support;
- Add end-user instructions on how to download and install SabreDAV;
- Remove bundled log4php code due to licensing issues;
- Remove included SabreDAV 1.6.x version;
- Upgrade to SabreDAV 1.8 API.

* Wed Apr 17 2013 Alfred Klomp <alfred@bokxing-it.nl> 0.17.1
- Add SabreDAV license file, to comply with SabreDAV license terms.


config.inc.php
<?php
/*
 * Copyright 2011 - 2012 Guillaume Lapierre
 * Copyright 2012 - 2013 Bokxing IT, http://www.bokxing-it.nl
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3, 
 * as published by the Free Software Foundation.
 *  
 * "Zarafa" is a registered trademark of Zarafa B.V. 
 *
 * This software use SabreDAV, an open source software distributed
 * with New BSD License. Please see <http://code.google.com/p/sabredav/>
 * for more information about SabreDAV
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *  
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Project page: <http://github.com/1afa/sabre-zarafa/>
 * 
 */

	// Configuration file for Zarafa SabreDAV interface

	// Location of the SabreDAV server.
	// see SabreDAV documentation: 
	//      http://code.google.com/p/sabredav/wiki/CardDAV#Get_clients_working
	define ('CARDDAV_ROOT_URI', '/sabre-zarafa');

	// Zarafa server location
	define ('ZARAFA_SERVER', 'file:///var/run/zarafa');
	
	// Authentication realm
	define ('SABRE_AUTH_REALM', 'Zarafa SabreDAV CardDav');
	
	// Product ID sent in vcards
	define ('VCARD_PRODUCT_ID', '-//SabreDav/ZarafaBackend/' . SABRE_ZARAFA_VERSION);
	
	// Choose VCard version
	// Supported values:
	// - 2 : old 2.1 format
	// - 3 : 3.0 format - compatible with OS.X addressbook
	// - 4 : newer 4.0 format - compatible with emClient
	define('VCARD_VERSION', 3);
	
	// Pattern to generate the "name" of the contact in Zarafa
	// Only used when no SORT-AS or X-CN is provided
	// unless SAVE_AS_OVERRIDE_SORTAS is set to true
	//
	// If only a company name is given, use it
	//
	// Options available:
	// %d - display name (default up to 0.12)
	// %l - last name
	// %f - first name
	// %c - company name
	define ('SAVE_AS_PATTERN', '%d');
	define ('SAVE_AS_OVERRIDE_SORTAS', false);
	
	// Charset to convert data to.
	// iPhone does not support UTF8 nor windows contact
	// vcard 4 are supposed to be utf8 encoded according to RFC :(
	// This is a ICONV parameter so one can use //TRANSLIT if needed
	// Default empty: let producer decide! (vcard4 utf8, other: iso-8859-1)
	define ('VCARD_CHARSET', '');
		// utf-8
		// 'ISO-8859-1//TRANSLIT');
	
	// If set to true all the write operations will be refused
	define ('READ_ONLY', false);

	// Allow SabreDAV browser plugin
	define('SABRE_DAV_BROWSER_PLUGIN', true);
	
	// Some clients do not add a TYPE= attribute for telephone numbers
	// This parameters maps empty TYPE to a MAPI attribute
	// unmapped properties are not exported by sabre-zarafa to VCARDs
	// unless used as DEFAULT_TELEPHONE_NUMBER_PROPERTY.
	
	// Valid values are:
	//  home_telephone_number			(mapped)
	//	home2_telephone_number			(mapped)
	//	cellular_telephone_number		(mapped)
	//	office_telephone_number			(mapped)
	//	business2_telephone_number		(mapped)
	//	business_fax_number				(mapped)
	//	home_fax_number					(mapped)
	//	pager_telephone_number			(mapped)
	//	isdn_number						(mapped)
	//	company_telephone_number		(mapped)
	//	car_telephone_number			(mapped)
	//	assistant_telephone_number		(mapped)
	//	primary_telephone_number		(mapped)
	//	callback_telephone_number		* unmapped *
	//	other_telephone_number			* unmapped *
	//	primary_fax_number				* unmapped *
	//	radio_telephone_number			* unmapped *
	//	telex_telephone_number			* unmapped *
	//	ttytdd_telephone_number			* unmapped *
	define ('DEFAULT_TELEPHONE_NUMBER_PROPERTY', 'primary_telephone_number');
	
	// If set, missing information from VCard will be removed from existing contact
	// if present. This could be lead to information loss if CardDav client
	// does not handle correctly all information or does not send back some
	// information.
	// Information sent as X-PROPERTY are not deleted even if this setting
	// is set to true
	define ('CLEAR_MISSING_PROPERTIES', true);
	
	// If set to false, one cannot delete a collection
	// As SabreDav does not handle sub-folders, deleting "root" collection
	// would delete all folders and contacts which might be dangerous
	define ('ALLOW_DELETE_FOLDER', true);

	// If set to true, all shared address books from other
	// users are included under the own account. The list is taken
	// from set Zarafa Webapp setting. You will need to make sure use a
	// rename pattern the includes the provenance if there are name
	// conflicts with own address books
	define ('INCLUDE_SHARED_ADDRESSBOOKS', false);
	
	// When set to true SAVE_RAW_VCARD, the vCard will not
	// only be parsed but also saved as a custom property to the contact.
	// When such a vCard is attached to a contact this vCard is sent
	// back to the requesting client instead of being generated from
	// mapi properties. This should be a performance boost but it will
	// require some storage capacity.
	// This should improve compatibility with CardDav clients using fields
	// that do not map easily with zarafa (multiple IMPP in emClient for
	// example)
	define ('SAVE_RAW_VCARD', false);
	
	// How to "write" dates to VCard
	define ('DATE_PATTERN', 'Ymd');
	
	// VCard cache version. Change cache version to force generation of
	// vcards from contact properties
	define ('CACHE_VERSION', '2.' . VCARD_VERSION);
	
	// Set default timezone
	if (function_exists("date_default_timezone_set")) {
		date_default_timezone_set('Europe/Paris');
	}

	// Return ETags to the client. Normally, with reasonable clients like
	// OSX Contacts or SoGo Connector, you want this. However, according to
	// the SabreDAV documentation, there are certain clients (which?) that
	// become confused when, for a given ETag, we don't return
	// byte-for-byte the exact body that the client sent. So support for
	// this feature is kept optional:
	define('ETAG_ENABLE', true);

	// Change the name of the folder as returned by Sabre-Zarafa.
	// Format options available:
	//  %d - display name (default up to 0.20)
	//  %p - provenance, either 'public' or 'private' or the other user's
	//       name if shared
	define('FOLDER_RENAME_PATTERN', '%d');
